File: kiosk/src/lib/sync.ts
/**
 * Downloading the move's data into the kiosk's local database, and the
 * status store the UI watches while it happens.
 *
 * `runSync(initiativeId, initiativeName)` fetches both sync endpoints in
 * parallel and only then writes `assets` and `people` — a fetch that
 * fails leaves the previously cached rows exactly as they were, so a
 * kiosk that loses the network keeps the move it already has. The meta
 * `sync` row records which move, the two counts, and when, which is also
 * how a reloaded kiosk shows `done` with real counts before anything is
 * fetched again (`hydrateSyncStatus`).
 *
 * The store/hook shape mirrors `setupState.ts` (`useSyncExternalStore`
 * over a module-level snapshot + listener set). Sync outcomes never
 * touch `kiosk_setup_complete`: a kiosk is set up whether or not its
 * local copy downloaded.
 */

import { useEffect, useSyncExternalStore } from 'react';

import { ApiError, fetchAssetsSync, fetchPeopleSync } from './api';
import { readMeta, replaceAll, writeMeta } from './localDb';

export type SyncPhase = 'idle' | 'running' | 'done' | 'error';

export interface SyncStatus {
  phase: SyncPhase;
  assets?: number;
  people?: number;
  syncedAt?: string;
  error?: string;
}

const META_KEY = 'sync';

let status: SyncStatus = { phase: 'idle' };
const listeners = new Set<() => void>();

function setStatus(next: SyncStatus): void {
  status = next;
  listeners.forEach((fn) => fn());
}

export function readSyncStatus(): SyncStatus {
  return status;
}

export function subscribeSyncStatus(fn: () => void): () => void {
  listeners.add(fn);
  return () => listeners.delete(fn);
}

/** Back to `idle` with no counts — the Developer tab's "Clear local
 *  data" pairs this with `clearDb()`. */
export function resetSyncStatus(): void {
  hydrated = false;
  setStatus({ phase: 'idle' });
}

let hydrated = false;

/** Reads the persisted meta row once per page load so a reload shows
 *  `done` + the counts without re-fetching. Never disturbs a sync that
 *  is already running or finished in this session. */
export async function hydrateSyncStatus(): Promise<void> {
  if (hydrated) return;
  hydrated = true;
  try {
    const row = await readMeta(META_KEY);
    if (!row || status.phase !== 'idle') return;
    setStatus({
      phase: 'done',
      assets: typeof row.assets === 'number' ? row.assets : 0,
      people: typeof row.people === 'number' ? row.people : 0,
      syncedAt: typeof row.syncedAt === 'string' ? row.syncedAt : undefined,
    });
  } catch {
    /* no local database yet (or storage refused): stay idle */
  }
}

/**
 * Fetch both endpoints, then replace both stores. Failure modes:
 * a fetch error reports the `ApiError` code, a database error reports
 * `'storage'`; either way the cached rows are untouched and the status
 * keeps the counts it had so the footer doesn't blink.
 */
export async function runSync(initiativeId: string, initiativeName: string): Promise<void> {
  const previous = status;
  setStatus({ ...previous, phase: 'running', error: undefined });

  let assets: Awaited<ReturnType<typeof fetchAssetsSync>>;
  let people: Awaited<ReturnType<typeof fetchPeopleSync>>;
  try {
    [assets, people] = await Promise.all([fetchAssetsSync(initiativeId), fetchPeopleSync()]);
  } catch (err) {
    setStatus({
      ...previous,
      phase: 'error',
      error: err instanceof ApiError ? err.code : 'unknown_error',
    });
    return;
  }

  try {
    await replaceAll('assets', assets.assets);
    await replaceAll('people', people.people);
    const syncedAt = new Date().toISOString();
    await writeMeta(META_KEY, {
      initiativeId, initiativeName,
      assets: assets.assets.length, people: people.people.length, syncedAt,
    });
    hydrated = true;
    setStatus({
      phase: 'done', assets: assets.assets.length, people: people.people.length, syncedAt,
    });
  } catch {
    setStatus({ ...previous, phase: 'error', error: 'storage' });
  }
}

/** The status snapshot, hydrating from the meta row on first mount. */
export function useSyncStatus(): SyncStatus {
  const snapshot = useSyncExternalStore(subscribeSyncStatus, readSyncStatus);
  useEffect(() => { void hydrateSyncStatus(); }, []);
  return snapshot;
}

/** "synced 2:14 PM" — the time alone; the kiosk syncs per shift, so a
 *  date would be noise on the summary card. */
export function formatSyncedAt(iso: string): string {
  return new Date(iso).toLocaleTimeString(undefined, { hour: 'numeric', minute: '2-digit' });
}

File: Android_Kiosk_App/app/src/main/java/com/serversherpa/kiosk/data/sync/Sync.kt
package com.serversherpa.kiosk.data.sync

import androidx.room.withTransaction
import com.serversherpa.kiosk.core.ApiError
import com.serversherpa.kiosk.data.api.KioskApi
import com.serversherpa.kiosk.data.api.KioskJson
import com.serversherpa.kiosk.data.db.KioskDatabase
import com.serversherpa.kiosk.data.db.MOVE_META_KEY
import com.serversherpa.kiosk.data.db.MetaEntity
import com.serversherpa.kiosk.data.db.toEntity
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable

enum class SyncPhase { IDLE, RUNNING, DONE, ERROR }

data class SyncStatus(
    val phase: SyncPhase = SyncPhase.IDLE,
    val assets: Int? = null, val people: Int? = null, val containers: Int? = null, val trucks: Int? = null,
    val syncedAt: String? = null, val error: String? = null,
)

@Serializable
data class SyncMeta(
    val initiativeId: String, val initiativeName: String,
    val assets: Int, val people: Int, val containers: Int, val trucks: Int, val syncedAt: String,
)

/**
 * kiosk/src/lib/sync.ts: fetch all four endpoints in parallel, then
 * replace all four tables and the meta row in ONE transaction. A failed
 * fetch leaves the cached rows untouched. Sync never touches setup state.
 */
class Sync(
    private val api: KioskApi,
    private val db: KioskDatabase,
    private val scope: CoroutineScope,
    private val clock: () -> Long = System::currentTimeMillis,
) {
    private val _status = MutableStateFlow(SyncStatus())
    val status: StateFlow<SyncStatus> = _status
    private var currentRun = 0
    private var hydrated = false

    /** Reads the persisted meta row once so a relaunch shows DONE with counts. */
    suspend fun hydrate() {
        if (hydrated) return
        hydrated = true
        val meta = readMeta(db) ?: return
        if (_status.value.phase != SyncPhase.IDLE) return
        _status.value = SyncStatus(SyncPhase.DONE, meta.assets, meta.people, meta.containers, meta.trucks, meta.syncedAt)
    }

    fun run(initiativeId: String, initiativeName: String) { scope.launch { runNow(initiativeId, initiativeName) } }

    suspend fun runNow(initiativeId: String, initiativeName: String) {
        val myRun = ++currentRun
        val previous = _status.value
        _status.value = previous.copy(phase = SyncPhase.RUNNING, error = null)
        val fetched = try {
            coroutineScope {
                val a = async { api.syncAssets(initiativeId) }
                val p = async { api.syncPeople() }
                val c = async { api.syncContainers(initiativeId) }
                val t = async { api.syncTrucks(initiativeId) }
                Fetched(a.await(), p.await(), c.await(), t.await())
            }
        } catch (e: Exception) {
            if (myRun != currentRun) return
            _status.value = previous.copy(phase = SyncPhase.ERROR, error = if (e is ApiError) e.code else "unknown_error")
            return
        }
        if (myRun != currentRun) return
        try {
            val syncedAt = Instant.ofEpochMilli(clock()).toString()
            db.withTransaction {
                db.assets().deleteAll(); db.assets().insertAll(fetched.assets.assets.map { it.toEntity() })
                db.people().deleteAll(); db.people().insertAll(fetched.people.people.map { it.toEntity() })
                db.containers().deleteAll(); db.containers().insertAll(fetched.containers.containers.map { it.toEntity() })
                db.trucks().deleteAll(); db.trucks().insertAll(fetched.trucks.trucks.map { it.toEntity() })
                val meta = SyncMeta(
                    initiativeId, initiativeName, fetched.assets.assets.size, fetched.people.people.size,
                    fetched.containers.containers.size, fetched.trucks.trucks.size, syncedAt,
                )
                db.meta().put(MetaEntity(MOVE_META_KEY, KioskJson.encodeToString(SyncMeta.serializer(), meta)))
            }
            if (myRun != currentRun) return
            hydrated = true
            _status.value = SyncStatus(
                SyncPhase.DONE, db.assets().count(), db.people().count(), db.containers().count(), db.trucks().count(), syncedAt,
            )
        } catch (e: Exception) {
            if (myRun != currentRun) return
            _status.value = previous.copy(phase = SyncPhase.ERROR, error = "storage")
        }
    }

    /** "Clear local data": the move tables and the meta row — never the outbox. */
    suspend fun clearLocalData() {
        db.withTransaction {
            db.assets().deleteAll(); db.people().deleteAll(); db.containers().deleteAll(); db.trucks().deleteAll()
            db.meta().delete(MOVE_META_KEY)
        }
        hydrated = false
        _status.value = SyncStatus()
    }

    private class Fetched(
        val assets: com.serversherpa.kiosk.core.model.KioskAssetsSync,
        val people: com.serversherpa.kiosk.core.model.KioskPeopleSync,
        val containers: com.serversherpa.kiosk.core.model.KioskContainersSync,
        val trucks: com.serversherpa.kiosk.core.model.KioskTrucksSync,
    )

    companion object {
        suspend fun readMeta(db: KioskDatabase): SyncMeta? = db.meta().get(MOVE_META_KEY)?.let {
            try { KioskJson.decodeFromString(SyncMeta.serializer(), it.value) } catch (e: Exception) { null }
        }

        /** "2:14 PM" — the time alone; the kiosk syncs per shift. */
        fun formatSyncedAt(iso: String): String = try {
            DateTimeFormatter.ofPattern("h:mm a").format(Instant.parse(iso).atZone(ZoneId.systemDefault()))
        } catch (e: Exception) { iso }
    }
}
